tic_tac_toe/board.py

```diff
@@ -0,0 +1,48 @@
+from tic_tac_toe.views import BoardDisplay
+
+
+class Board:
+    def __init__(self):
+        self.board = [" ", " ", " ", " ", " ", " ", " ", " ", " "]
+
+    def __getitem__(self, index):
+        return self.board[index]
+
+    def turn(self):
+        view = BoardDisplay()
```

Comment:
The `Board` class is starting to look more generic to any kind of Tic-Tac-Toe game, but your `BoardDisplay` class is very specific to playing in a terminal on the command line...

Reply:
@palledorous I know you know this concept because you did it on the Roman Numeral project! Does this line ring any bells?
https://github.com/palledorous/roman-numeral-converter-project/pull/6/files#diff-60b949dcd8ab2109f89d8d77c1f53e75R6

Reply:
(Nodding) exactly, only as with elsewhere, you don't need to define a default value for the arg. The method is perfectly fine saying "in order to call me, you need to pass in a [converter|display], and I'm not going to make any assumptions about which one (of potentially many different kinds) you want me to use!"

Reply:
Cool, I think I'm starting to get it.

Reply:
Something like this?

```
class Board(BoardDisplayInterface):
...
```

Reply:
That would be some sort of inheritance thing—your board class would be a child of the BoardDisplayInterface. You're making it tougher than it needs to be!
```python
class Board:
    def turn(self, presenter):
        # do some things
        presenter.display_board(self.board)
        # do more things
```
With that change, you could pass in _any_ object as the second argument to `turn`, and as long as it has a `display_board` method on it, it'll Just Work :tm:. The `Board` class (specifically its `turn` method) depends on an _interface_.

Then, "higher up" somewhere, something can pass specific _implementations_ of the interface, for example:
```python
import BlackAndWhiteCommandLinePresenter
import ColorCommandLinePresenter
import Board

def run():
    board = Board()
    view_preference = get_view_preference_from_user() # some prompt and input, maybe?
    if view_preferences = "COLOR":
        presenter = ColorCommandLinePresenter()
    else:
        presenter = BlackAndWhiteCommandLinePresenter()

    while True:
        board.turn(presenter)
```

Reply:
NVM, I got it. So it's just a matter of instantiating the Board class with a view attribute like so:
```
class Board:
    def __init__(self, view):
        self.board = [" ", " ", " ", " ", " ", " ", " ", " ", " "]
        self.view = view
```
This way, in the `turn()` method I can call `self.view.display_board(self, self.board)`

Because in my `run()` method that controls the game, I pass in `CommandLineBoardPresenter` when I instantiate the Board.

```
def run():
    board = Board(CommandLineBoardPresenter)
```

Reply:
Implied in my snippet above is that the two classes (`Color...` and `B/W...`) both have a method defined named `display_board`, but those two methods do different things.

This can be a tough concept to grasp in a dynamic language because interfaces are all implicit. They'll become clearer in statically typed languages.

Reply:
Yes, you've got it. You're right that that could be either on the `turn` method or on the `init` method and kept around as an attribute of each particular instance. Your way is probably the more OO approach, so go with that for now!

Reply:
It looks like we were both commenting at the same time! I wish I'd seen yours first. Thank you for the help! It's much clearer now.